diff --git a/arch/Kconfig b/arch/Kconfig
index 0ecbd6d42..6fc60b732 100644
--- a/arch/Kconfig
+++ b/arch/Kconfig
@@ -855,4 +855,8 @@ config VMAP_STACK
 	  the stack to map directly to the KASAN shadow map using a formula
 	  that is incorrect if the stack is in vmalloc space.
 
+# Select if arch has all set_memory_ro/rw/x/nx() functions in asm/cacheflush.h
+config ARCH_HAS_SET_MEMORY
+	bool
+
 source "kernel/gcov/Kconfig"
diff --git a/include/linux/sched.h b/include/linux/sched.h
index 768183eec..95aac9211 100755
--- a/include/linux/sched.h
+++ b/include/linux/sched.h
@@ -2207,6 +2207,9 @@ struct task_struct {
 	u64 susfs_last_fake_mnt_id;
 #endif
 	struct thread_struct thread;
+#ifdef CONFIG_SECURITY
+	void				*security;
+#endif
 /*
  * WARNING: on x86, 'thread_struct' contains a variable-sized
  * structure.  It *MUST* be at the end of 'task_struct'.
